### ISSUE TITLE
Jupyter Notebooks should be activating selected environment upon changing kernel

### ISSUE THREAD

**Issue by yilmazdurmaz** [microsoft/vscode-jupyter#12918]
EDIT: the next two post has logs and screenshots of the current behavior I am having.

Type: <b>Feature Request</b>

Some time ago, I was playing with the notebooks but got some problems I could not solve and give up using them from within vscode.

Recently, I was trying to work with R kernel in a jupyter notebook. I wanted to give vscode another go on this. kernels, extensions, and language servers. all are fine, except the environment.

Changing a kernel was seem to be very easy for python, and also I thought this kernel selection was activating the virtual environment the kernel belongs. now I see the problem with how it works: no environment activated.

I checked this twice; one without an active environment set for R kernel, and the other where vscode started from within a terminal with an active environment with R kernel ready. the second one runs fine while the first gives `Failed to start the Kernel. The kernel died.` Also, R extension cannot find R installation and stays `not attached`.

In addition, new internal terminals are started without no environment activated. This is also another sign that kernels in these environments are put into use without respecting their installed environments.

In short, please do not just try to directly use kernels (and executables) found in an environment folder, but instead activate their corresponding environments first so as to access other libraries installed within those environments.

VS Code version: Code 1.75.1 (441438abd1ac652551dbe4d408dfcec8a499b8bf, 2023-02-08T21:32:34.589Z)
OS version: Windows_NT x64 10.0.19044
Modes:
Sandboxed: No

**Comment by yilmazdurmaz** [comment]
Hi @DonJayamanne,

I believe you reached this "duplicate" opinion just because I used "terminal" keyword in just one paragraph. Unfortunately, the terminal is just a minor problem and it seems only for PowerShell as I do not let it run scripts. although it has its own problem, "cmd.exe" can start with a "conda activate R", so in short terminal is not the main problem here.

I have checked the other post. The logs they show are way different than what I got here. I wasn't sure what to check and what to add here. Now I see your "verbose" output request there, tried to come up few more explanative and repetitive results from my setup.

I included a test file with 3 cells to test the 3 occurrences in the screenshot below. If I select an item from the "env" list (shown as suggested from the kernel list view), then the kernel is working. If I select python or R kernel from the "kernel" list, there is a long list of logs you may inspect, all ending up with failed kernel start-ups. trying to select a python kernel in this R environment, 

Interestingly, for the python, I am getting "unknown option --slave" because, for some reason, this kernel is being tried to be activated with the parameters from the R kernel.

test jupyter file: [vscode-jupyter-kernel-problem.zip](https://github.com/microsoft/vscode-jupyter/files/10818212/vscode-jupyter-kernel-problem.zip)
logs: [vscode-jupyter-kernel-log.txt](https://github.com/microsoft/vscode-jupyter/files/10818179/vscode-jupyter-kernel-log.txt)

python selected in the env list
![select python in the env list](https://user-images.githubusercontent.com/1812105/221023912-e687c0ef-4526-4ccc-9f88-f8368b3a30f5.png)
python selected in the kernel list
![select python in the kernel list](https://user-images.githubusercontent.com/1812105/221023936-5c5bd003-e974-4685-8b10-1e1c98649402.png)
R selected in the kernel list
![select R in the kernel list](https://user-images.githubusercontent.com/1812105/221023948-a39f23a0-3916-404d-acb7-c9fa1e1252d6.png)

PS: I will add one more post that shows what happens if I start the vscode from a terminal that has my R environment activated. it is just that writing takes a bit of time.

**Comment by yilmazdurmaz** [comment]
Now for the mostly "working" part.

for this purpose, I followed this routine and the remaining parts are similar.
- open a terminal
- activate the environment I will work `conda activate R` (be aware that both irkernel and ipykernel are installed)
- run vscode from this terminal `code .`

This time, open clicking on the kernel selection button, there is a slight difference to what I am getting: env list now includes R executable, but it is removed from the kernel list.

![Screenshot (16)](https://user-images.githubusercontent.com/1812105/221029910-c8caeafd-0b86-4b8c-b617-c2b17d825dd0.png)

this log follows the same order, python from env list and python kernel list, but R is now from env list. R kernel is working just fine when I start vscode this way.

[vscode-jupyter-kernel-log2.txt](https://github.com/microsoft/vscode-jupyter/files/10818545/vscode-jupyter-kernel-log2.txt)

I hope you can find the problem (and also see the difference between the other post)

PS: You may be referring to them by some other names, but I have used "env list" and "kernel list" because of what I see. the list is separated so that "conda env" is in the above list (hence env list), and "jupyter kernel" is in the list below (hence kernel list).

**Comment by rebornix** [comment]
From what I read above, this issue mentioned two:

* activate env when picking a non-Python kernel
* "In addition, new internal terminals are started without no environment activated", which is https://github.com/microsoft/vscode-jupyter/issues/12452

We didn't seem to have a solution for the first one yet unless I missed something.

**Comment by yilmazdurmaz** [comment]
Hi @rebornix ,
- the terminal issue on my side was possibly my fault for not giving permission to PowerShell to execute scripts. However, I cannot remember what I was experiencing at the time I wrote the first post. however, although "cmd.exe" has a hiccup, it eventually runs a command that activates the environment, so it is safe to assume my problem with it was not an issue at all. I cannot say about that other linked issue though. I could not understand what they mean by "internal terminal", but sounds different than my issue here.
- my problem is not just about a "non-python kernel" because selecting a python kernel "in the kernel list" also causes an error.
  - selecting an environment in "Conda Env" list (upper list) activates the environment and kernels run fine if their executable is listed here.
  - selecting a kernel in "Jupyter Kernel" list (bottom list) seems to try to activate an environment, but fails I guess because there are errors and no kernel works with this selection. (all the red arrows and "selected" keywords in the screenshots)
  - If I start vscode from a terminal with an environment activated (the second one above), non-python kernel is removed from "jupyter kernel" list, put into "conde env" list, and works fine between switches. but this "run from an activated terminal" part should not be a required step here.
 
The log files I attached above include all the outputs when I switch/activate environments/kernels. The lines are too long and I have no idea what to look for to identify anything useful to ease things for you. I may make a video of the problem if the info I provided is confusing.

**Comment by DonJayamanne** [comment]
Let’s keep things simple, is the Problem that toy cannot run code against the R kernel?

**Comment by yilmazdurmaz** [comment]
yes and no. 
- R kernel works when I start vscode from an environment-activated terminal.
- R kernel does not work if I start vscode directly and try to activate R kernel later, because it fails to activate the kernel.
- it is not only about R kernel. python kernel also does not activate when I select it from the "kernel list". also gives this "--slave" error that is confused with parameters from R kernel.

**Comment by DonJayamanne** [comment]
Duplicate of https://github.com/microsoft/vscode-jupyter/issues/11519

The issue with python Envs should be fixed in the pre-release version
The pair with R will be fixed in the next release, see #11519

**Comment by yilmazdurmaz** [comment]
@DonJayamanne, I peeked at the post you linked. at first glance, the main issue in it is the detection of kernels. it is not the case for me. I will check it thoroughly and let you know in case I find the two are different.
Thanks.
